Framework/PythonInterface/plugins/algorithms/FindMultipleUMatrices.py

```diff
@@ -0,0 +1,380 @@
+        return out_props[0] if len(out_props) == 1 else out_props
+
+    def remove_equiv_ubs_by_accuracy(self, hkl_ers, ubs, pointgroup, unit_cell, min_angle_ub):
+        iub_keep = set()
+        for iub, ub in enumerate(ubs):
+            if iub in iub_keep:
+                continue
+            iub_similar = [
+                int(iub_other)
+                for iub_other in range(len(ubs))
+                if self.are_ubs_equivalent_within_tolerance(ub, ubs[iub_other], pointgroup, unit_cell, min_angle_ub) and iub_other != iub
+            ]
+            if iub_similar:
+                ibest, *_ = self.get_indices_of_nub_most_accurate_ubs(hkl_ers[tuple(iub_similar), :], num_ubs=1)
+                iub_keep.add(iub_similar[next(iter(ibest))])
+            else:
+                iub_keep.add(iub)
+        iub_keep = list(iub_keep)
+        return hkl_ers[iub_keep], [ubs[ii] for ii in iub_keep]
+
+    @staticmethod
+    def are_ubs_equivalent_within_tolerance(ub1, ub2, pointgroup, unit_cell, angle_tol):
+        inv_ub1 = np.linalg.inv(ub1)
+        inv_ub2 = np.linalg.inv(ub2)
+        axes_equiv = 2 * [False]
+        # loop over hkl vectors along 2 of QLab cartesian axes
+        for iqax in range(2):
+            # loop over all equivalent HKL
+            for hkl1 in pointgroup.getEquivalents(inv_ub1[iqax, :]):
+                if abs(unit_cell.recAngle(*hkl1, *inv_ub2[iqax, :], 1)) < angle_tol:
+                    axes_equiv[iqax] = True
+                    break
+            if not axes_equiv[iqax]:
+                break
+        return all(axes_equiv)
+
+    def get_indices_of_nub_most_accurate_ubs(self, hkl_ers, num_ubs=1):
+        # see which ubs index the most peaks most accurately (with minimum error)
+        iub_min_er = np.argmin(hkl_ers, axis=0)
+        # exclude reflections that are not indexed by any UB
+        hkl_ers_indexed = np.isfinite(hkl_ers)
+        ipks_indexed = np.any(hkl_ers_indexed, axis=0)
+        iub, nindex_best = np.unique(iub_min_er[ipks_indexed], return_counts=True)
+        # remove ubs that don't index any peaks most accurately
+        ikeep = nindex_best > 0
+        iub = iub[ikeep]
+        nindex_best = nindex_best[ikeep]
+        # sort by ubs that index peaks most accurately
+        # if same num peaks indexed most accurately then additionally by total number peaks indexed
+        nindex_total = hkl_ers_indexed[iub, :].sum(axis=1)  # total num peaks indexed by each UB within tol
+        isort = np.lexsort((-nindex_total, -nindex_best))
+        iub = iub[isort].astype(int)
+        if len(iub) <= num_ubs:
+            return iub, iub_min_er, ipks_indexed
+        else:
+            hkl_ers[iub[-1], :] = np.inf  # overwrite values as if didn't index any peaks
+            return self.get_indices_of_nub_most_accurate_ubs(hkl_ers, num_ubs)
+
+    def get_peak_tables_for_each_ub(self, peaks, ubs, hkl_ers, num_ubs, optimise_ubs):
```

Comment:
Would you be happy to add some type hints for the methods and staticmethods in this file? It might help with readability in future

Reply:
Yes, sorry I'm bad at doing type hinting from the beginning!